src/TurbulenceClosures/turbulence_closure_implementations/scale_invariant_smagorinsky.jl

```diff
@@ -0,0 +1,331 @@
+
+function with_tracers(tracers, closure::ScaleInvariantSmagorinsky{TD, FT}) where {TD, FT}
+    Pr = tracer_diffusivities(tracers, closure.Pr)
+    return ScaleInvariantSmagorinsky{TD, FT}(closure.C, closure.averaging, Pr)
+end
+
+function LᵢⱼMᵢⱼ_ccc(i, j, k, grid, u, v, w)
+    return (      L₁₁ᶜᶜᶜ(i, j, k, grid, u, v, w) * M₁₁ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1)
+            +     L₂₂ᶜᶜᶜ(i, j, k, grid, u, v, w) * M₂₂ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1)
+            +     L₃₃ᶜᶜᶜ(i, j, k, grid, u, v, w) * M₃₃ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1)
+            + 2 * L₁₂ᶜᶜᶜ(i, j, k, grid, u, v, w) * M₁₂ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1) 
+            + 2 * L₁₃ᶜᶜᶜ(i, j, k, grid, u, v, w) * M₁₃ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1) 
+            + 2 * L₂₃ᶜᶜᶜ(i, j, k, grid, u, v, w) * M₂₃ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1) )
+end
+
+function MᵢⱼMᵢⱼ_ccc(i, j, k, grid, u, v, w)
+    return (      M₁₁ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1)^2
+            +     M₂₂ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1)^2
+            +     M₃₃ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1)^2
+            + 2 * M₁₂ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1)^2
+            + 2 * M₁₃ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1)^2
+            + 2 * M₂₃ᶜᶜᶜ(i, j, k, grid, u, v, w, 2, 1)^2)
+end
+
+
+@kernel function compute_LM_MM!(LM, MM, grid, closure, velocities)
+    i, j, k = @index(Global, NTuple)
+    @inbounds LM[i, j, k] = LᵢⱼMᵢⱼ_ccc(i, j, k, grid, velocities...)
+    @inbounds MM[i, j, k] = MᵢⱼMᵢⱼ_ccc(i, j, k, grid, velocities...)
+end
+
+@kernel function _compute_scale_invariant_smagorinsky_viscosity!(νₑ, LM_avg, MM_avg, grid, closure, buoyancy, velocities, tracers)
+    i, j, k = @index(Global, NTuple)
+
+    # Strain tensor dot product
+    Σ² = ΣᵢⱼΣᵢⱼᶜᶜᶜ(i, j, k, grid, velocities.u, velocities.v, velocities.w)
+
+    cₛ² = max(LM_avg[i, j, k] / MM_avg[i, j, k], 0)
+    @inbounds νₑ[i, j, k] = cₛ² * (Δᶠ(i, j, k, grid))^2 * sqrt(2Σ²)
+end
+
+@inline directional_average!(averaged_field, field,   ::XDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=1)))
+@inline directional_average!(averaged_field, field,   ::YDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=2)))
+@inline directional_average!(averaged_field, field,   ::ZDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=3)))
+@inline directional_average!(averaged_field, field,  ::XYDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=(1, 2))))
+@inline directional_average!(averaged_field, field,  ::XZDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=(1, 3))))
+@inline directional_average!(averaged_field, field,  ::YZDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=(2, 3))))
+@inline directional_average!(averaged_field, field, ::XYZDirectionalAveraging) = averaged_field .= compute!(Field(Average(field, dims=(1, 2, 3))))
```

Comment:
This part definitely needs to be changed. I just haven't been able to come up an appropriate kernel that works.

Reply:
What kernel? If you want to compute an average then you shouldn't write your own kernel. You probably want to preallocate the result though. The code will be a lot simpler since then you just call `compute!`.